src/Qir/Runtime/public/QirRuntime.hpp

```diff
@@ -16,140 +16,141 @@ extern "C"
     // ------------------------------------------------------------------------
 
     // Allocate one qubit. Qubit is guaranteed to be in |0> state.
-    // Qubit needs to be released via quantum__rt__qubit_release.
-    QIR_SHARED_API QUBIT* quantum__rt__qubit_allocate(); // NOLINT
+    // Qubit needs to be released via __quantum__rt__qubit_release.
+    QIR_SHARED_API QUBIT* __quantum__rt__qubit_allocate(); // NOLINT
```

Comment:
I believe there can be a more optimal sequence of changes.

Let's consider by example of qubit allocation.
Before this PR we had 2 functions:
* `quantum__rt__qubit_allocate()` implemented in C/C++, I will call it _C implementation_ below;
* `__quantum__rt__qubit_allocate()` implemented in IR, I will call it _IR implementation_ below.

Both are exposed externally. The IR implementation calls the C implementation. 

The Q# compiler generates the IR files that call the IR implementation.
The other language environments (like Python, Rust) capable of using C interface, can call the C implementation directly.
Eventually we want to get rid of the IR implementation and get the Q# compiler to generate the IR that calls the C implementation directly too.

To achieve that, we can make a switch in the Q# compiler from calling the IR implementation to calling the C implementation _first_. And then (when nobody will be calling the IR implementation any more) just to remove the IR implementation. These two steps are done in different repos, that is why they are easy to do. I feel that this is the most optimal sequence of changes.

Now about the changes in this PR.
This PR removes the IR implementation and gives the C implementation the name starting with double underscore `__`. Q#-compiler-generated code still calls the `__` names (`__quantum__rt__qubit_allocate()`). I assume that _this is an intermediate step_, and the next step will be to make a change in two repos in parallel:
* in the Q# compiler to call `quantum__rt__qubit_allocate()` instead of `__quantum__rt__qubit_allocate()`,
* in the qsharp-runtime to rename `__quantum__rt__qubit_allocate()` to `quantum__rt__qubit_allocate()`.

If that's the case (this is an intermediate step) then this approach seems to me less optimal because of the several reasons:
* the identifiers starting with `__` (reserved in the C/C++ standard) are present in the C/C++ code, which is undesirable, at least some of the community may not like it (and the other community members' compilers or static analyzers will complain);
* the next step (parallel renaming in two repos) can be postponed for an indefinite period of time, and we can go public with `__` names;
* the parallel change in two repos is somewhat tricky.

If that's NOT the case (this is NOT an intermediate step) then the first of the reasons (some of the community may not like it) will stay (until the parallel change in the two repos).

Reply:
I think you make some great points here, and it's worth having more discussion about which QIR spec updates we make and how those get reflected into the compiler and the runtime. But I think the question of how long to keep these .ll files and how we would remove them is tangential to all of this.  Here's my two cents...

> The Q# compiler generates the IR files that call the IR implementation.
The other language environments (like Python, Rust) capable of using C interface, can call the C implementation directly.
Eventually we want to get rid of the IR implementation and get the Q# compiler to generate the IR that calls the C implementation directly too.

I don't think this part is correct. We don't have a goal to write in other languages something that calls into the C implementation directly, we have a goal to robustly implement the QIR spec so that any QIR program can be compiled against our runtime regardless of where it was generated from. The QIR spec defines an API, and the Q# compiler produces QIR that uses that API. Then any runtime that wants to support QIR implements that API. IR vs C is an implementation detail of our particular runtime and not something that affects any other runtime someone would want to use. And if someone uses a different front end language, like Python or Rust, to write a quantum program they want to compile into QIR then they will need to use whatever API is defined in the QIR spec, not what our runtime happens to implement in C.

Let's take the same runtime functions you highlighted for qubit allocation:

> `quantum__rt__qubit_allocate()` implemented in C/C++, I will call it C implementation below;
`__quantum__rt__qubit_allocate()` implemented in IR, I will call it IR implementation below.

The second function, `__quantum__rt__qubit_allocate()` matches the API defined in the [QIR spec](https://github.com/microsoft/qsharp-language/blob/main/Specifications/QIR/Quantum-Runtime.md#qubits), such that any QIR program that is compliant with the spec can include calls to this function and would expect it to return a valid qubit. The first function, `quantum__rt__qubit_allocate()` doesn't match the name defined in the spec, so would not be considered valid QIR. It's just an implementation detail of our runtime, which has the second function forward directly to the first:
https://github.com/microsoft/qsharp-runtime/blob/d32e2890688a04cb4bca4436deeefba314bc540a/src/Qir/Runtime/lib/QIR/bridge-rt.ll#L151-L155
Except in the special case of `%Range` none of these forwards provide any real value. The bridge-rt.ll file is not distributed separately such that others can build on it, so it is not an extensibility point. The fact that the C implementation functions are extern is a necessity in order to be able to refer to them in the .ll files but not intended to make them available for use outside of our runtime. So removing this extra layer and defining the C implementation to match exactly the signatures from the QIR spec means we are more directly creating an implementation of QIR API.

I agree with you that we should update the QIR spec to make friendlier names, but I'm not so sure anymore about how to treat underscores specifically. Looking at the [cppreference page on identifiers](https://en.cppreference.com/w/cpp/language/identifiers), it actually states that "identifiers with a double underscore anywhere are reserved" so it's not just the leading underscores that are a problem. But that is a separate discussion that we should pursue along with cleaning up our runtime implementation. I would assert that whatever names and signatures the spec defines we should try to implement as extern C functions, and if that is not possible (like in the case of  passing `%Range` by value), then we should implement it as best we can and file issues against the spec for update. And once the spec does get updated, only then would we change the API surface implemented by the runtime.

You are right that such a spec update would be tricky and require multi-repo changes. I think the pattern should be
1. Update the QIR spec in the language repo.
2. Provide updated API implementation in the runtime repo alongside of the old one.
3. Update the Q# compiler to produce QIR compliant with the new spec and implementation.
4. Remove the old API implementation in the runtime repo when we are confident it should no longer be used.

This change doesn't affect that plan because it doesn't change the QIR API surface we implement, it just implements it in C instead of in IR forwarders.

All that being said, there is no rush to get this change in, especially because it only affects internal implementation details. We can take the time to make sure the team agrees with this and/or adjust the approach before merging it.

Reply:
Stefan, thank you very much for a detailed explanation! It has pointed to me that the importance of the QIR Spec has fallen out of my attention, which resulted in some misunderstanding on my side.
Now I clearly understand the reasons behind the changes in this PR and totally agree with each item that you provide in your explanation (including the fragment where you write "I don't think this part is correct" ;-) ).
Your 4 steps to the new API are the right thing to do, as opposed to what I previously considered as "more optimal sequence of changes".
I bookmarked this explanation in the description of the internal deliverable "Refactor QIR RT to expose C interface".